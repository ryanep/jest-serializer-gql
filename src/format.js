import { buildASTSchema, printSchema } from 'graphql';

export default (schema, options = {}) => {
  if (options.ast) {
    return JSON.stringify(schema, null, 2);
  }

  const ast = buildASTSchema(schema);
  return printSchema(ast).trim();
};
