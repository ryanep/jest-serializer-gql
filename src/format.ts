import { ASTNode } from 'graphql';
import { print } from 'graphql/language/printer';

export const format = (node: ASTNode): string => {
  const output = print(node).trim();
  return output;
};
