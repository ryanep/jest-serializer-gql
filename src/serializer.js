import createSerializer from './create-serializer';

export default createSerializer();
