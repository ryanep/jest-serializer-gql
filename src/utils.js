export const isGraphqlSchema = schema => {
  if (
    typeof schema === 'object' &&
    schema.kind &&
    schema.definitions &&
    schema.loc
  ) {
    return true;
  }
  return false;
};
