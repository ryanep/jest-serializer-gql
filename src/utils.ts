import { ASTNode } from 'graphql';

export const isGraphqlSchema = (node: ASTNode): boolean => {
  if (node && typeof node === 'object' && node.kind && node.loc) {
    return true;
  }
  return false;
};
