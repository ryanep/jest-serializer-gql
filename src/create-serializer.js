import { isGraphqlSchema } from './utils';
import format from './format';

export default function createSerializer(options) {
  return {
    test(val) {
      return isGraphqlSchema(val);
    },
    print(val) {
      return format(val, options);
    },
  };
}
