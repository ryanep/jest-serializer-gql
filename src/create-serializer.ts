import { ASTNode } from 'graphql';
import { isGraphqlSchema } from './utils';
import { format } from './format';

export const createSerializer = (): jest.SnapshotSerializerPlugin => ({
  test(node: ASTNode): boolean {
    return isGraphqlSchema(node);
  },
  print(node: ASTNode): string {
    return format(node);
  },
});
