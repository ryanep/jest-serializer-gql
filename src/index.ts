import { createSerializer } from './create-serializer';

export { createSerializer };
