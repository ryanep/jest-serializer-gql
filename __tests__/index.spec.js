import gql from 'graphql-tag';
import { createSerializer } from '../src';

const schema = `
	type User {
    id: ID!
    firstName: String! @deprecated
	}

	type Query {
		users(firstName: String): [User]
	}

	type Mutation {
		login(emailAddress: String! password: String!): User
	}
`;

describe('graphql schema', () => {
  it('should match snapshot when exporting AST', () => {
    expect.addSnapshotSerializer(createSerializer({ ast: true }));
    const ast = gql(schema);
    expect(ast).toMatchSnapshot();
  });

  it('should match GraphQL schema snapshot with default settings', () => {
    expect.addSnapshotSerializer(createSerializer());
    const ast = gql(schema);
    expect(ast).toMatchSnapshot();
  });
});
