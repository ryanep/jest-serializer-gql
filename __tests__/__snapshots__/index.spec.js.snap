// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`graphql schema should match GraphQL schema snapshot with default settings 1`] = `
type Mutation {
  login(emailAddress: String!, password: String!): User
}

type Query {
  users(firstName: String): [User]
}

type User {
  id: ID!
  firstName: String! @deprecated
}
`;

exports[`graphql schema should match snapshot when exporting AST 1`] = `
{
  "kind": "Document",
  "definitions": [
    {
      "kind": "ObjectTypeDefinition",
      "name": {
        "kind": "Name",
        "value": "User"
      },
      "interfaces": [],
      "directives": [],
      "fields": [
        {
          "kind": "FieldDefinition",
          "name": {
            "kind": "Name",
            "value": "id"
          },
          "arguments": [],
          "type": {
            "kind": "NonNullType",
            "type": {
              "kind": "NamedType",
              "name": {
                "kind": "Name",
                "value": "ID"
              }
            }
          },
          "directives": []
        },
        {
          "kind": "FieldDefinition",
          "name": {
            "kind": "Name",
            "value": "firstName"
          },
          "arguments": [],
          "type": {
            "kind": "NonNullType",
            "type": {
              "kind": "NamedType",
              "name": {
                "kind": "Name",
                "value": "String"
              }
            }
          },
          "directives": [
            {
              "kind": "Directive",
              "name": {
                "kind": "Name",
                "value": "deprecated"
              },
              "arguments": []
            }
          ]
        }
      ]
    },
    {
      "kind": "ObjectTypeDefinition",
      "name": {
        "kind": "Name",
        "value": "Query"
      },
      "interfaces": [],
      "directives": [],
      "fields": [
        {
          "kind": "FieldDefinition",
          "name": {
            "kind": "Name",
            "value": "users"
          },
          "arguments": [
            {
              "kind": "InputValueDefinition",
              "name": {
                "kind": "Name",
                "value": "firstName"
              },
              "type": {
                "kind": "NamedType",
                "name": {
                  "kind": "Name",
                  "value": "String"
                }
              },
              "directives": []
            }
          ],
          "type": {
            "kind": "ListType",
            "type": {
              "kind": "NamedType",
              "name": {
                "kind": "Name",
                "value": "User"
              }
            }
          },
          "directives": []
        }
      ]
    },
    {
      "kind": "ObjectTypeDefinition",
      "name": {
        "kind": "Name",
        "value": "Mutation"
      },
      "interfaces": [],
      "directives": [],
      "fields": [
        {
          "kind": "FieldDefinition",
          "name": {
            "kind": "Name",
            "value": "login"
          },
          "arguments": [
            {
              "kind": "InputValueDefinition",
              "name": {
                "kind": "Name",
                "value": "emailAddress"
              },
              "type": {
                "kind": "NonNullType",
                "type": {
                  "kind": "NamedType",
                  "name": {
                    "kind": "Name",
                    "value": "String"
                  }
                }
              },
              "directives": []
            },
            {
              "kind": "InputValueDefinition",
              "name": {
                "kind": "Name",
                "value": "password"
              },
              "type": {
                "kind": "NonNullType",
                "type": {
                  "kind": "NamedType",
                  "name": {
                    "kind": "Name",
                    "value": "String"
                  }
                }
              },
              "directives": []
            }
          ],
          "type": {
            "kind": "NamedType",
            "name": {
              "kind": "Name",
              "value": "User"
            }
          },
          "directives": []
        }
      ]
    }
  ],
  "loc": {
    "start": 0,
    "end": 193
  }
}
`;
